src/domain/core/src/utils/time_source.rs

```diff
@@ -24,46 +28,132 @@ pub trait SystemTimeSource: Send + Sync {
 #[dill::interface(dyn SystemTimeSource)]
 pub struct SystemTimeSourceDefault;
 
+#[async_trait::async_trait]
 impl SystemTimeSource for SystemTimeSourceDefault {
     fn now(&self) -> DateTime<Utc> {
         Utc::now()
     }
+
+    async fn wake_after(&self, duration: Duration) {
+        let std_duration = duration.to_std().unwrap();
+
+        tokio::time::sleep(std_duration).await;
+    }
 }
 
 /////////////////////////////////////////////////////////////////////////////////////////
 
+type AwaitingCallerId = usize;
+type WokeUpCallersIds = Vec<AwaitingCallerId>;
+
+#[derive(Debug)]
+struct AwaitingCaller {
+    id: AwaitingCallerId,
+    time_to_wake_up: Duration,
+    waker_tx: Option<oneshot::Sender<()>>,
+}
+
+#[derive(Debug, Default)]
+struct SystemTimeSourceStubState {
+    t: Option<DateTime<Utc>>,
+    next_caller_id: AwaitingCallerId,
+    awaiting_callers: Vec<AwaitingCaller>,
+}
+
 #[derive(Debug, Clone)]
 pub struct SystemTimeSourceStub {
-    t: Arc<Mutex<Option<DateTime<Utc>>>>,
+    state: Arc<Mutex<SystemTimeSourceStubState>>,
 }
 
 impl SystemTimeSourceStub {
     pub fn new() -> Self {
         Self {
-            t: Arc::new(Mutex::new(None)),
+            state: Arc::new(Mutex::new(SystemTimeSourceStubState::default())),
```

Comment:
I have big doubts that with the logic we are introducing it's correct to allow `None` for `t`.

Reply:
and it's very right that you have -- it was a mistake to leave it as `Optional` 

fixed